domain/utxoindex/utxoindex.go

```diff
@@ -20,143 +17,143 @@ type UTXOIndex struct {
 }
 
 // New creates a new UTXO index
-func New(consensus externalapi.Consensus, database database.Database) *UTXOIndex {
-	store := newUTXOIndexStore(database)
-	return &UTXOIndex{
+func New(consensus externalapi.Consensus, database database.Database) (*UTXOIndex, error) {
+	utxoIndex := &UTXOIndex{
 		consensus: consensus,
-		store:     store,
+		store:     newUTXOIndexStore(database),
 	}
-}
-
-// Update updates the UTXO index with the given DAG selected parent chain changes
-func (ui *UTXOIndex) Update(chainChanges *externalapi.SelectedChainPath) (*UTXOChanges, error) {
-	onEnd := logger.LogAndMeasureExecutionTime(log, "UTXOIndex.Update")
-	defer onEnd()
 
-	ui.mutex.Lock()
-	defer ui.mutex.Unlock()
+	isSynced, err := utxoIndex.isSynced()
+	if err != nil {
+		return nil, err
+	}
 
-	log.Tracef("Updating UTXO index with chainChanges: %+v", chainChanges)
-	for _, removedBlockHash := range chainChanges.Removed {
-		err := ui.removeBlock(removedBlockHash)
+	if !isSynced {
+		err = utxoIndex.Reset()
 		if err != nil {
 			return nil, err
 		}
 	}
-	for _, addedBlockHash := range chainChanges.Added {
-		err := ui.addBlock(addedBlockHash)
+
+	return utxoIndex, nil
+}
+
+// Reset deletes the whole UTXO index and resyncs it from consensus.
+func (ui *UTXOIndex) Reset() error {
+	err := ui.store.deleteAll()
+	if err != nil {
+		return err
+	}
+
+	virtualInfo, err := ui.consensus.GetVirtualInfo()
+	if err != nil {
+		return err
+	}
+
+	var fromOutpoint *externalapi.DomainOutpoint
+	for {
+		const step = 1000
+		virtualUTXOs, err := ui.consensus.GetVirtualUTXOs(virtualInfo.ParentHashes, fromOutpoint, step)
 		if err != nil {
-			return nil, err
+			return err
```

Comment:
Do we want to check the error and restart the loop if the parents changed? or do we assert here that the flows are off and if they're not then we want to crash here?

Reply:
Yeah, I think I prefer to assert